pennylane/gradients/gradient_transform.py

```diff
@@ -144,8 +164,32 @@ def jacobian_wrapper(*args, **kwargs):
                 # is present inside the QNode.
                 return qjac
 
-            # Classical processing of a single argument is present. Return qjac @ cjac.
-            jac = qml.math.squeeze(qml.math.tensordot(qml.math.T(cjac), qjac, [[-1], [-1]]))
-            return qml.math.T(jac)
+            # Classical processing present of either:
+            #   a) a single argument
+            #   b) multiple arguments of the same shape
+            # The shape of the classical jacobian returned by qml.jacobian depends on the scenario:
+            #   a) (# gate args, qnode arg shape)
+            #   b) (reverse qnode arg shape, # gate args, # qnode args)
+            # It then needs to be contracted with the quantum jacobian of shape:
+            #      (qnode output shape, # gate args)
+            # The result should have shape:
+            #   a) (qnode ouput shape, qnode arg shape)
+            #   b) (reverse qnode arg shape, qnode output shape, # qnode args)
+            num_gate_args = qml.math.shape(qjac)[-1]
+            trainable_args_idx = self._jacobian_trainable_args(args, qnode.interface)
+            qnode_arg_shape = qml.math.shape(args[trainable_args_idx[0]])
+            num_qnode_args = len(trainable_args_idx)  # need trainable number of qnode args only
+
+            if qml.math.shape(cjac) == (num_gate_args, *qnode_arg_shape):
+                # single QNode argument
+                jac = qml.math.tensordot(qjac, cjac, [[-1], [0]])

```

Comment:
For my understanding: this is the case that worked with the previous version of the function, correct? And it was refactored.

Reply:
I've just added the old behaviour back in proposal commit 913ba9b, it's a bit a different as it had to work for both branches that are present now (although incorrectly mind you). But yes I believe the single QNode argument case always worked previously.

I would like to keep the old behaviour as a fallback just in case, even though I can't conceive of a situation where it would be invoked. But given that we will modify how qml.jacobian works in the near future, I think it would be ok to leave it in as a defensive move until then. (I originally removed it to satisfy codecov but this commits adds it back with `pragma: no cover`.)

What do you think?

Reply:
What would give the impression that the `else` branch is required? Not quite sure it makes sense to add it if there's no specific example that would be expected to hit the branch.

Reply:
You are right, the else branch really should not be required, that is in theory it should always be either condition 1:
`qml.math.shape(cjac) == (num_gate_args, *qnode_arg_shape)`
or condition 2:
`qml.math.shape(cjac) == (*qnode_arg_shape[::-1], num_gate_args, num_qnode_args)`.
If it were to be executed it would be because there is a bug. I just feel uneasy because condition 1 & 2 do not cover the entire range of the data type.

I guess we could either:
- return a default value (e.g. empty Jacobian)
  ```py
  jac = ()
  if qml.math.shape(cjac) == (num_gate_args, *qnode_arg_shape):
      ...
  elif qml.math.shape(cjac) == (*qnode_arg_shape[::-1], num_gate_args, num_qnode_args):
      ...
  ```
- return a default with a warning
  ```py
  if qml.math.shape(cjac) == (num_gate_args, *qnode_arg_shape):
      ...
  elif qml.math.shape(cjac) == (*qnode_arg_shape[::-1], num_gate_args, num_qnode_args):
      ...
  else:
      jac = ()
      warnings.warn("Could not compute hybrid jacobian...")
  ```
- just assume condition 2 is true whenever condition 1 is false
  ```py
  if qml.math.shape(cjac) == (num_gate_args, *qnode_arg_shape):
      ...
  else:   # qml.math.shape(cjac) == (*qnode_arg_shape[::-1], num_gate_args, num_qnode_args)
      # multiple QNode arguments with stacking
      cjac = qml.math.swapaxes(cjac, -1, -2)
      jac = qml.math.tensordot(cjac, qjac, [[-1], [-1]])
      jac = qml.math.moveaxis(jac, -1, len(qnode_arg_shape))
  ```
- assume but assert condition 2 is true whenever condition 1 is false
  ```py
  if qml.math.shape(cjac) == (num_gate_args, *qnode_arg_shape):
      ...
  else:
      assert qml.math.shape(cjac) == (*qnode_arg_shape[::-1], num_gate_args, num_qnode_args)
      # multiple QNode arguments with stacking
      cjac = qml.math.swapaxes(cjac, -1, -2)
      jac = qml.math.tensordot(cjac, qjac, [[-1], [-1]])
      jac = qml.math.moveaxis(jac, -1, len(qnode_arg_shape))
  ```
- raise an error
  ```py
  if qml.math.shape(cjac) == (num_gate_args, *qnode_arg_shape):
      ...
  elif qml.math.shape(cjac) == (*qnode_arg_shape[::-1], num_gate_args, num_qnode_args):
      ...
  else:
      raise ValueError ...
  ```
  
  Maybe this question is more about the general programming philosophy 😅

Reply:
To me, having a branch with a warning (or even error) would sound the best.

We expect cases to fall into the first two categories. If that's not the case, then something unexpected is going on. I think it makes sense to try a way of performing the computation, but warn the user about it. 👍